### DISCUSSION TITLE
Fixed-point number formatting

### DISCUSSION THREAD

**Discussion by mbait** [modm-io/modm#1064] [Ideas]
Would you be interested in accepting a code for formatting fixed-point numbers as strings? I saw that you have a plan to integrate the fpm library some day, but for a basic needs, say an IIR filter implementation, it's enough to have a multiplication procedure and a number-to-text conversion procedure - addition and subtraction is same (if saturation is not needed) as for integers. This is the prototype that I currently use in my project https://godbolt.org/z/99dG9K31r.

**Comment by salkinium** [comment]
I also wrote [a fpm type](https://github.com/salkinium/upainter/blob/master/ges/math/fixed_point.hpp) some time ago, however, there were many corner cases that would've required a lot more carefully written unit tests. Instead I would very much prefer to integrate an existing, tested library to reduce the maintenance burden. You can of course add the small implementation externally in application code, or in a fork of modm. But in general, I'd really like to delegate as much as possible to not reinvent the wheel so much…

**Reply by mbait** [reply]
That's a good point, but looking at the fpm library the first thing that comes to mind is that the code isn't really optimized for embedded. What do you think?

**Reply by salkinium** [reply]
Hm, it seems fine, we cannot integrate the `ios.hpp`, since we don't support the stdlib iostreams, but the rest should be fine. Do you want to try it out in your application first and see if it works for your use-case?

**Reply by mbait** [reply]
No really as for now my primary target platform is AVR, so all work with fixed-point types is narrowed down to using fmul/fmuls instructions along with wring a custom fixed-point-to-string conversion routine. The latter is for debug output only though, so a full-scale fixed-point math library seems to be an overkill in my case, but I'm interested in that subject in general, so will definitely give fpm a try some day.

**Reply by salkinium** [reply]
You can use `scons listing` to show the disassembly of the code to validate what instructions the compiler generates. That often helps me understand what overhead the C++ code has.
